codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java

```diff
@@ -0,0 +1,156 @@
+            return decompressor.status();
+        } catch (Exception e) {
+            handleException(e);
+            throw e;
+        }
+    }
+
+    private boolean downstreamWantsMore(ChannelHandlerContext ctx) {
+        return downstreamWantsMore || ctx.channel().config().isAutoRead();
+    }
+
+    private void processSome(ChannelHandlerContext ctx) {
+        while (!closed && downstreamWantsMore(ctx)) {
+            Decompressor.Status status = status();
+            switch (status) {
+                case NEED_OUTPUT:
+                    ByteBuf buf;
+                    try {
+                        buf = decompressor.takeOutput();
+                    } catch (Exception e) {
+                        handleException(e);
+                        throw e;
+                    }
+                    if (!buf.isReadable()) {
+                        // try again.
+                        buf.release();
+                        break;
+                    }
+                    downstreamWantsMore = false;
+                    ctx.fireChannelRead(buf);
+                    ctx.fireChannelReadComplete();
+                    break;
+                case NEED_INPUT:
+                    if (inputBuffer == null) {
+                        return;
+                    }
+                    Object item = inputBuffer.remove(0);
+                    if (inputBuffer.isEmpty()) {
+                        inputBuffer.recycle();
+                        inputBuffer = null;
+                    }
+                    try {
+                        decompressor.addInput((ByteBuf) item);
+                    } catch (Exception e) {
+                        handleException(e);
+                        throw e;
+                    }
+                    break;
+                case COMPLETE:
+                    decompressor.close();
+                    closed = true;
+                    break;
+                default:
+                    throw new AssertionError("Unknown status: " + status);
+            }
+        }
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        reading = true;
+        if (closed) {
+            ReferenceCountUtil.release(msg);
+            return;
+        }
+        if (status() == Decompressor.Status.NEED_INPUT) {
+            try {
+                decompressor.addInput((ByteBuf) msg);
+            } catch (Exception e) {
+                handleException(e);
+                throw e;
+            }
+        } else {
+            if (inputBuffer == null) {
+                inputBuffer = RecyclableArrayList.newInstance();
+            }
+            inputBuffer.add(msg);
+        }
+        processSome(ctx);
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        reading = false;
+        if (!ctx.channel().config().isAutoRead() && downstreamWantsMore) {
+            ctx.read();
+        }
```

Comment:
Like mentioned above we need to do something more smart about `channelReadComplete` and so might need to also handle it here.